### ISSUE TITLE
Shooting snapshot not in initial trajectory?

### ISSUE THREAD

**Issue by dwhswenson** [choderalab/openpathsampling#456]
Quick background: the analysis in #454 is intended to work with any set of MC steps that include shooting movers. The committor calculation is just one example of that; but a similar shooting point analysis can be useful in lots of cases, so I'm trying to make it work in general.

I'm getting some weird results on occasion from normal TIS calculations. It popped up while I was developing the committor analysis, but here I've given the simplest explanation I can to explain what is going wrong.

First, I run the local tests to ensure I have an updated `mstis.nc` file. Then I do the following:

```python
import openpathsampling as paths
storage = paths.AnalysisStorage("mstis.nc")

for step in storage.steps:
    try:
        # if not a shooting/extension move, this will raise an exception
        shooting_snap = step.change.canonical.trials[0].details.shooting_snapshot
    except AttributeError:
        pass # ignore the moves that aren't shooting
    except IndexError:
        pass # ignore the moves that aren't shooting
    else:
        trial_traj = step.change.canonical.trials[0].trajectory
        init_traj = step.change.canonical.trials[0].details.initial_trajectory
        if not trial_traj.is_correlated(init_traj):
            print step.change
```

For any (one-way) shooting move, the trial trajectory should be correlated with the initial trajectory. However, in a recent test, this gives the following output:

```
PathSimulatorStep : PathSampling : Step # 66 with 0 samples
 +- RandomChoice :
 |   +- RandomChoice :
 |   |   +- OneWayShooting :
 |   |   |   +- SampleMove : BackwardShootMover : False :[]
PathSimulatorStep : PathSampling : Step # 99 with 1 samples
 +- RandomChoice :
 |   +- RandomChoice :
 |   |   +- OneWayShooting :
 |   |   |   +- SampleMove : ForwardShootMover : True : 1 samples [<Sample @ 0x10f8a18d0>]
```

Okay... that's weird. So let's look at those steps:

```python
step66 = storage.steps[66]
step99 = storage.steps[99]

print step66.change.canonical.trials[0].ensemble == step99.change.canonical.trials[0].ensemble
print step99.change.canonical.trials[0].ensemble.name
```

```
True
I'face 2
```

Same ensemble, but not a special one (i.e., not MS-outer, so that's not the cause.) Let's look some more at each of those:

```python
stateA = storage.volumes["A"]
stateB = storage.volumes["B"]
stateC = storage.volumes["C"]
vol_names = {state.name: state for state in [stateA, stateB, stateC]}
```

```python
step66_trial = step66.change.canonical.trials[0]
print step66_trial.details.initial_trajectory.summarize_by_volumes_str(vol_names)
print step66_trial.trajectory.summarize_by_volumes_str(vol_names)
print step66.change.accepted
```

```
A-None-B
B-None-B
False
```

```python
step99_trial = step66.change.canonical.trials[0]
print step99_trial.details.initial_trajectory.summarize_by_volumes_str(vol_names)
print step99_trial.trajectory.summarize_by_volumes_str(vol_names)
print step99.change.accepted
```

```
A-None-B
A-None-A
True
```

So, my guess is the problem of losing the snapshot identities has something to do with the trajectory segment changing states. This is relatively unusual (hence the reason it only happens twice in this run, and not at all in some runs). Although I've only explored enough to see it happen 3 times (these 2 plus 1 before), 3/3 of the cases where I saw the correlation disappear involved changing state in the new segment of the trajectory.

I'm not sure how this is happening; I'll look into it more later this week, but I wanted to post this in case you had an idea as to what was going on, @jhprinz.

**Comment by jhprinz** [comment]
Extremely good catch. I have no idea how you managed to find this. But yes, that is a bug. I started looking into it yesterday with no clue, but had the right guess today. 

To be fair. It does not loose information, but in some rare cases it saves the same snapshot again which destroys the information necessary for correlation between trajectories. You would have to really check snapshots for equality and not rely on the same index in the storage file.

In short:

1. When you have a snapshot that has never been reversed 
2. so that it has no reversed snapshot created yet
3. then save it
4. then reverse it
5. and then save the reversed snapshot
6. the reversed snapshot is not registered with the storage (it can't) since it was not used with it yet.

So the solution is to test before saving if the reversed snapshot is already stored. This I did not implement and it is only a few lines of code.

I also added better handling of storing proxy objects. 

There is still a chance that you will store a snapshot twice which is in the following scenario:

1. load snapshot x from file1
2. save it in file2
3. clear the cache from file1 (or load enough other snapshots from file1 that snapshot x is not in the cache) and remove snapshot x from memory
4. load snapshot x again from file1 (this will create a new object in memory)
5. save it in file2 (this will store it again)

Problem is that we would have to track exactly which object in file1 would correspond to objects in file2 and I think the handling of that should be in the responsibility of the user. The overhead can be vast. The problem will disappear once we use UUIDs.

PR to fix this is coming soon!

**Comment by dwhswenson** [comment]
> Extremely good catch. I have no idea how you managed to find this.

Well, I'm glad **you** figured out what the problem was! I don't think I would have found that.

My part was because I knew the committor analysis could be useful when applied to regular shooting moves (imagine starting with a committor analysis, but then histogramming by some CV applied to the snapshots: this gives you a decent picture of the committor along that CV).

So I wrote the analysis in such a way that it would apply to any list of `steps`, and extract the ones with `shooting_snapshot`s associated (as in the example code). Then I wrote a little function to get the part of the trial trajectory that wasn't in the initial trajectory (basically, the opposite of `Trajectory.shared_snapshots`). That way, I can figure out which state is the one the trajectory shot *into*, instead of the part it already had. From there it wasn't long until I got runtime errors that eventually led me to discover that the shooting snapshot wasn't in the initial trajectory.

Anyway, very glad that you found the problem, and it sounds like it won't be a difficult fix. The remaining chance of failure is not one that I think I'm likely to encounter, and it sounds like a long-term fix is already on the way with the UUID stuff.
